# HG changeset patch
# User Tartalo
# Date 1278534136 -7200
# Branch trunk
# Node ID 8590d3da443039a427436a8e115ac3b2875f40ae
# Parent  f75327bb2fcd24a13b1e338e0077db99f6f0aca2
Deny access to instances if other party/user with the same perm id is inside
Fixes #2844

diff -r f75327bb2fcd -r 8590d3da4430 src/server/game/Maps/Map.cpp
--- a/src/server/game/Maps/Map.cpp	Wed Jul 07 10:42:42 2010 +0200
+++ b/src/server/game/Maps/Map.cpp	Wed Jul 07 22:22:16 2010 +0200
@@ -2395,6 +2395,32 @@
         return false;
     }
 
+    // cannot enter if instance is in use by another party/soloer that have a
+    // permanent save in the same instance id
+
+    PlayerList const &playerList = GetPlayers();
+    Player *firstInsidePlayer = NULL;
+
+    if (!playerList.isEmpty())
+        for (PlayerList::const_iterator i = playerList.begin(); i != playerList.end(); ++i)
+            if (Player *iPlayer = i->getSource())
+            {
+                if (iPlayer->isGameMaster()) // bypass GMs
+                    continue;
+                if (!player->GetGroup()) // player has not group and there is someone inside, deny entry
+                {
+                    player->SendTransferAborted(GetId(), TRANSFER_ABORT_MAX_PLAYERS);
+                    return false;
+                }
+                // player inside instance has no group or his groups is different to entering player's one, deny entry
+                if (!iPlayer->GetGroup() || iPlayer->GetGroup() != player->GetGroup() )
+                {
+                    player->SendTransferAborted(GetId(), TRANSFER_ABORT_MAX_PLAYERS);
+                    return false;
+                }
+                break;
+            }
+
     return Map::CanEnter(player);
 }
 
